### PR TITLE
Add support for de Bruijn indices

### PR DESCRIPTION
**Pull request by michaelpj**
This seems useful to have, for tests if nothing else. I haven't implemented using it for serialization, as suggested in https://github.com/input-output-hk/plutus/issues/471, but it's there if we want it.

- Add de Bruijn indices as kinds of name.
- Adds transformations between terms with uniques and terms with de Bruijn indices.
    - Also check that these round-trip. (We have to use the "interesting" term generator, since we need the terms to be well-scoped, so we can't use the simple term generator.)
- Debruijnify terms before printing them in tests, making them more robust to changes.
    - I noticed that the PIR-producing tests aren't producing debug output, which they should be, will fix later. Although maybe I'll need to write another debruijnifier for PIR...

### CODE REVIEW

File: language-plutus-core/src/Language/PlutusCore/DeBruijn.hs [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+
+instance HasIx (TyDeBruijn a) where
+    index = _Wrapped' . index
+
+-- Converting from normal names to DeBruijn indices, and vice versa
+
+{- Note [Levels and indices]
+The indices that we use for de Bruijn identifiers are *relative*. However,
+when doing conversions it is easier to work with *absolute* levels, since  that
+way we don't have to adjust them all when we go under a binder.
```
Comment (effectfully) [inline language-plutus-core/src/Language/PlutusCore/DeBruijn.hs]: I misread "easier to work with *absolute* levels" as "easier to return *absolute* levels" as opposed to "easier to have *absolute* levels in state, even though we return *relative* indices". Maybe state explicitly that a term with indices is returned? You describe calculations below from which it follows that indices are returned, but I think explicit words would be helpful here.

File: language-plutus-core/examples/Language/PlutusCore/Examples/Data/InterList.hs [hunk 2]
```diff
@@ -7,8 +7,11 @@ module Language.PlutusCore.Examples.Data.InterList
     , getBuiltinFoldrInterList
     ) where
 
-import           Language.PlutusCore
 import           Language.PlutusCore.MkPlc
+import           Language.PlutusCore.Name
+import           Language.PlutusCore.Quote
+import           Language.PlutusCore.Renamer
+import           Language.PlutusCore.Type
```
Comment (effectfully) [inline language-plutus-core/examples/Language/PlutusCore/Examples/Data/InterList.hs]: I do not have anything against this change, but I wonder why you made it.

File: language-plutus-core/src/Language/PlutusCore/DeBruijn.hs [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+import qualified Data.Bimap                 as BM
+import qualified Data.Text                  as T
+import           Data.Typeable
+
+import           Numeric.Natural
+
+import           GHC.Generics
+
+-- | A relative index used for de Bruijn identifiers.
+newtype Ix = Ix Natural
```
Comment (effectfully) [inline language-plutus-core/src/Language/PlutusCore/DeBruijn.hs]: Not wild about stealing names from `base`.

File: language-plutus-core/src/Language/PlutusCore/DeBruijn.hs (hunk 1 above)
Comment (michaelpj) [reply language-plutus-core/src/Language/PlutusCore/DeBruijn.hs]: Yes, I'll make that clearer.

File: language-plutus-core/examples/Language/PlutusCore/Examples/Data/InterList.hs (hunk 2 above)
Comment (michaelpj) [reply language-plutus-core/examples/Language/PlutusCore/Examples/Data/InterList.hs]: I was doing something earlier that gave me an import cycle, but I don't need it now. OTOH I think importing the "everything" module inside `language-plutus-core` is somewhat bad practice anyway.

File: language-plutus-core/src/Language/PlutusCore/DeBruijn.hs (hunk 3 above)
Comment (michaelpj) [reply language-plutus-core/src/Language/PlutusCore/DeBruijn.hs]: Huh, I didn't know there was an `Ix` in base. I'll use `Index` which doesn't seem to be claimed by anything important.

**Pull request merged**